### PR TITLE
systemd hardening

### PR DESCRIPTION
**Pull request by candrews**
**Description of the Change**

Improve security by isolating boinc as much as possible.

See https://www.freedesktop.org/software/systemd/man/systemd.exec.html for explanations of each option
**Alternate Designs**

I don't think there are other designs; this is a means by which security can be improved. With this change, if a malicious job is sent to the boinc client, the damage it can do is significantly reduced.
**Release Notes**

N/A

### CODE REVIEW

**Comment by JuhaSointusalo** [comment]
How much testing you have done with this?

**Comment by candrews** [comment]
I've run it on my 2 systems for about a day now. One system is running Gentoo ~amd64 and has an rx580, it has been getting CPU and GPU tasks and hasn't encountered any problems. The other system runs Fedora 29 with an iGPU, it's also been working without trouble.

**Comment by JuhaSointusalo** [comment]
Ok, that's AMD and Intel GPUs. I think Nvidia GPU and VirtualBox apps need to be tested too. Also LHC's native Atlas app because that depends on a ton of stuff on the host.

Graphics apps should be tested too, that they can still communicate with science apps.

The problem with tightening security is that we run all kinds of apps. Ones that have been designed for BOINC environment and those that have been designed for some other use and might not even be coded according to any best practices. The latter ones I'm worried about. They might accidentally break with tightened security.

**Comment by candrews** [comment]
How do we actually do that testing?

File: client/scripts/boinc-client.service.in [hunk 1]
```diff
@@ -5,6 +5,19 @@ After=network-online.target
...
+NoNewPrivileges=true
+ProtectKernelModules=true
+ProtectKernelTunables=true
+ProtectControlGroups=true
+RestrictRealtime=true
+RestrictAddressFamilies=AF_INET AF_INET6 AF_UNIX
+RestrictNamespaces=true
+PrivateUsers=true
+CapabilityBoundingSet=
+ProtectControlGroups=true
```
Comment (JuhaSointusalo) [inline client/scripts/boinc-client.service.in]: duplicate

Comment (candrews) [reply client/scripts/boinc-client.service.in]: Sorry about that. I just forced push the removal of that duplication.

**Comment by JuhaSointusalo** [comment]
> How do we actually do that testing?

I don't really have any better idea other than run many projects and apps. Trying to use some intuition to spot the apps that are somehow different from all the others and that should be tested.

**Comment by candrews** [comment]
@Yamakuzure can you please help with testing?

**Comment by Yamakuzure** [comment]
> @Yamakuzure can you please help with testing?

I will. My Notebook has a discrete Nvidia Quadro M1200, that I use using Bumblebee. Both OpenCL with the integrated Intel and CUDA with the discrete Nvidia used to work.

I will report back once I have some results, but I am afraid I can only start testing tonight.

**Comment by Yamakuzure** [comment]
@candrews @JuhaSointusalo I have to eat my words. I can't test this change, as I do not run systemd.
*sigh* This happens if you are in too much of a hurry to take your time to actually _read_ everything slow and carefully.

I am sorry, but I do not think i can be of any help here, unless someone has an idea how to simulate these constraints without systemd. :frowning_face:

**Comment by TheAspens** [comment]
@lfield - would you be able to help with testing this - at least for LHC@Home?  @Germano0  - would you be able to help test/review this?  Thanks if either of you are able to help.

**Comment by Germano0** [comment]
@TheAspens yes I can help testing/reviewing the patch.
@candrews do you have a CUDA and/or OpenCL card? I would like to test if it can affect GPU detection too.
I should be able to use a nVidia card in next days for a limited time amount

**Comment by candrews** [comment]
> @candrews do you have a CUDA and/or OpenCL card? I would like to test if it can affect GPU detection too.

I have an RX580 (using OpenCL) and with this patch, GPU detection continues to work.

I don't have an nVidia/CUDA card.

**Comment by Germano0** [comment]
Sorry for having asked things you already wrote before.
I had some troubles this evening, so I am a bit absent-minded

**Comment by candrews** [comment]
No worries at all :-)

**Comment by Germano0** [comment]
@candrews actually on Fedora, the following flags do not permitt BOINC client to work properly. Next week I will better investigate on what happens
```
NoNewPrivileges=true
ProtectKernelModules=true
ProtectKernelTunables=true
RestrictRealtime=true
RestrictAddressFamilies=AF_INET AF_INET6 AF_UNIX
RestrictNamespaces=true

```

**Comment by Germano0** [comment]
@candrews yes you are completely right. I will fix them in SELinux

**Comment by Germano0** [comment]
- VirtualBox is not affected negatively by systemd hardening flags;
- GPU calculus is not affected negatively by systemd hardening flags. 

```
$ boinccmd --host localhost --passwd xxxx --get_tasks --active

======== Tasks ========
1) -----------
   name: LATeah2002L_1188.0_0_0.0_3562838_1
   WU name: LATeah2002L_1188.0_0_0.0_3562838
   project URL: https://einsteinathome.org/
   received: Mon Dec 17 10:21:04 2018
   report deadline: Mon Dec 31 10:21:03 2018
   ready to report: no
   state: downloaded
   scheduler state: scheduled
   active_task_state: EXECUTING
   app version num: 120
   resources: 1 CPU + 1 NVIDIA GPU
   estimated CPU time remaining: 1198.618292
   CPU time at last checkpoint: 420.989100
   current CPU time: 423.587700
   fraction done: 0.748597
   swap size: 14010 MB
   working set size: 195 MB
```

For me this pull request is perfectly fine

**Comment by candrews** [comment]
Happy New Year!

What's the latest news on this PR? I'm happy to do what I can to help.

**Comment by JuhaSointusalo** [comment]
Bad news. I started testing Atlas native app since it seemed nobody had tested that yet. First I ran the app with stock Debian 9 client to make sure the set up worked. Then I replaced the unit file with the one from this PR. Unfortunately it broke the app.

If the app is not run on Scientific Linux 6 it uses Singularity to spin up an SL6 container. Singularity needs root privileges to set up the container and with these hardening options that doesn't work. Singularity reports
```
tammi 07 20:07:31 debian9 Singularity[27146]: action-suid (U=118,P=27146)> Singularity is not running with appropriate privileges!
tammi 07 20:07:31 debian9 Singularity[27146]: action-suid (U=118,P=27146)> Check installation path is not mounted with 'nosuid', and/or consult manual.
tammi 07 20:07:31 debian9 Singularity[27146]: action-suid (U=118,P=27146)> Retval = 255
```
I tried disabling the obvious hardening options but that didn't help. Next I'll disable all of them and then them enable one by one.

**Comment by candrews** [comment]
How do you install the Atlas native boinc app?

It appears that setting up Atlas native is rather involved process requiring the installation of singularity and cvms. I'm using https://www.overclock.net/forum/18058-hardware-optimization/1699568-lhc-home-atlas-linux-native-app.html as my source (I can't find any terribly "officially" looking installation instructions)

If that's the case, then I suggest the Atlas native instructions be updated to include adding a systemd drop in file at `/etc/systemd/system/boinc-client.service.d/10-atlas_native.conf` changing whatever hardening settings need to be disabled when @JuhaSointusalo determines them.

The reason is that I think boinc should be secure by default - if there are a small subset of special case users that this security impacts, then let's handle them as a special case so we keep the vast majority of users as safe and secure as possible.

**Comment by JuhaSointusalo** [comment]
There is [a guide for building everything](https://lhcathome.cern.ch/lhcathome/forum_thread.php?id=4840) on LHC forums, though I skipped the building and installed CVMFS from [their repository](http://cernvm.cern.ch/portal/filesystem/downloads) and Singularity from [NeuroDebian](http://neuro.debian.net/pkgs/singularity-container.html). You'll need to create the `/scratch/cvmfs` directory and add `/etc/cvmfs/default.local` from the build instructions, I don't think they came with CVMFS packages.

I think the app is still in some sort of beta phase which might explain the lack of more official instructions.

**Comment by JuhaSointusalo** [comment]
This is what it looks like now:

```
# breaks setuid root
#ProtectKernelModules=true
#ProtectKernelTunables=true
#RestrictRealtime=true
#RestrictAddressFamilies=AF_INET AF_INET6 AF_UNIX
#PrivateUsers=true
#CapabilityBoundingSet=
#NoNewPrivileges=true

# ok
PrivateTmp=true
ProtectControlGroups=true
ProtectHome=true
ProtectSystem=strict
ReadWritePaths=/var/lib/boinc
RestrictNamespaces=true
```

I'll continue playing with it a while longer.

**Comment by JuhaSointusalo** [comment]
This didn't work for restoring setuid:
```
RestrictAddressFamilies=AF_INET AF_INET6 AF_UNIX
NoNewPrivileges=false
```

I guess the options are not applied in the order they are listed.

**Comment by JuhaSointusalo** [comment]
Ok, there is no way to allow new privileges when using any of the options that imply `NoNewPrivileges=true`.

https://github.com/systemd/systemd/blob/master/src/core/execute.c#L1394

My thinking was that those options+allow new privileges would have been better than nothing at all :(

I'll run one task with the client as it is now and another task with CVMFS cache cleared and after reboot so that there is nothing primed for the task. This will take two or three days.

**Comment by Germano0** [comment]
Before the holidays I was testing the flags one by one, before the end of
this week I will finish and fill a new comment with the results

**Comment by JuhaSointusalo** [comment]
I checked the packaging for Nvidia GPU drivers.

Nvidia's own package has only `nvidia-modprobe`, no udev rules. Arch Linux ships with pretty much the same files, though split into two packages. Debian has some udev files but also includes `nvidia-modprobe` so no idea how it really works. Ubuntu, RPM Fusion and Gentoo looks like they use udev rules.

We have some Arch users and we have some users who use the package from Nvidia. "Why the client needs to be restarted after X has loaded" used to be a frequent topic in forums. With the `NoNewPrivileges=true` that topic could come back.

File: client/scripts/boinc-client.service.in [hunk 2]
```diff
@@ -5,6 +5,18 @@ After=network-online.target
...
+NoNewPrivileges=true
+ProtectKernelModules=true
+ProtectKernelTunables=true
+ProtectControlGroups=true
+RestrictRealtime=true
+RestrictAddressFamilies=AF_INET AF_INET6 AF_UNIX
+RestrictNamespaces=true
+PrivateUsers=true
+CapabilityBoundingSet=
+ReadWritePaths=/var/lib/boinc
```
Comment (JuhaSointusalo) [inline client/scripts/boinc-client.service.in]: Needs to include `/etc/boinc-client`. Debian stores configuration files there and the client needs to be able to write to `cc_config.xml` and `global_prefs_override.xml`. (This means `gui_rpc_auth.cfg` and `remote_hosts.cfg` are protected only by filesystem permissions but that's the best we can do.)

**Comment by JuhaSointusalo** [comment]
The two Atlas native tasks have finally completed. They ran fine with the configuration that allows new privileges.

**Comment by Germano0** [comment]
```
ProtectHome=true
PrivateTmp=true
ProtectControlGroups=true
ReadWritePaths=/var/lib/boinc
```

are the only flags I managed to use with CPU calculus only on LHC@home. The others returns "calculus error" as working unit status.
Concerning GPU calculus, I am still running tests

**Comment by lfield** [comment]
@candrews Do you plan on making the suggested change?

**Comment by candrews** [comment]
I have made the suggested changes, thanks again!

**Comment by Germano0** [comment]
I am going  right now to test again `ProtectSystem=strict`, that in past led BOINC to run fail working units calculus. This time I will collect working units output in case of failure.

**Comment by Germano0** [comment]
I do confirm that flags

```
ProtectHome=true
PrivateTmp=true
ProtectSystem=strict
ProtectControlGroups=true
```

work good with
- "regular" CPU working units
- VirtualBox working units
- GPU working units
(tests runned on Fedora 29)

So in my humble opinion the pull request is okay

**Comment by lfield** [comment]
@JuhaSointusalo Please could you check to see if your concerns have been adddressed.

**Comment by candrews** [comment]
What's the latest news? I'd love to see this merged :)

**Comment by lfield** [comment]
I am happy to merge this but we are waiting for @JuhaSointusalo to address his review. If there is no activity by Thursday, it will be discussed at that the contributors call.

**Comment by Germano0** [comment]
For your info you will have to ask your Linux distribution selinux-policy maintainer to alllow
` ReadWritePaths=-/var/lib/boinc -/etc/boinc-client`
flag, otherwise you will not be able to start client service, like explained in this bugreport https://bugzilla.redhat.com/show_bug.cgi?id=1711682

**Pull request merged**